### DISCUSSION TITLE
Translate trimesh textures to pygfx

### DISCUSSION THREAD

**Discussion by schlegelp** [pygfx/pygfx#701] [Q&A] [answered]
Hi! 

I'm trying to write a converter that turns `trimesh` scenes (i.e. visuals + scene graph) into `pygfx` objects. Thankfully, `pygfx` makes most of this super easy and it's really just a few lines of boiler plate code. However, I'm currently a bit stuck when it comes to carrying over textures.

Here is a simple example to illustrate the issue. We'll be using the `fuze.obj` ([link](https://github.com/mikedh/trimesh/blob/main/models/fuze.obj)) + associated material files from the trimesh [texture example](https://trimesh.org/texture.html). 

This is what it ought to look like:

<img width="818" alt="Screenshot 2024-04-01 at 12 37 31" src="https://github.com/pygfx/pygfx/assets/7161148/da8a1113-4768-4d8d-91e2-2d81c4d41f4e">

This is what the trimesh material looks like:

```python
>>> import trimesh as tm
>>> mesh = tm.load('/Users/philipps/Downloads/fuze.obj')
>>> mesh 
<trimesh.Trimesh(vertices.shape=(664, 3), faces.shape=(1000, 3), name=`fuze.obj`)>
>>> # The image 
>>> mesh.visual.material.image
>>> # Which we can turn into a np array
>>> im = np.asarray(mesh.visual.material.image).astype(np.float32)
>>> im /= 255
>>> im.shape
(1024, 1024, 3)
>>> # The UV coordinates are here
>>> texcoords = mesh.visual.uv.astype(np.float32)
>>> texcoords.shape
(664, 2)
```

Here is how I tried to convert:

```python 
>>> import pygfx as gfx
>>> geometry = gfx.Geometry(
...            indices=mesh.faces.astype(np.int32, copy=False),
...            positions=mesh.vertices.astype(np.float32, copy=False)
...        )
>>> tex = gfx.Texture(im, dim=2)
>>> material = gfx.MeshPhongMaterial(map=tex)
>>> vis = gfx.Mesh(geometry, material)
>>> vis.geometry.texcoords = gfx.Buffer(texcoords)
```

Here is the result:

<img width="752" alt="Screenshot 2024-04-01 at 12 45 40" src="https://github.com/pygfx/pygfx/assets/7161148/58a80a89-8e39-49e8-827f-5495ecc74f48">

As you can see it did _something_ but it seems the UV coordinates are off. I tried flipping the axis in the image but that didn't do it either. It's probably an easy fix but as I said, I'm a bit stuck with this.

**Comment by Korijn** [comment, marked as answer]
There are some utilities to convert from trimesh to pygfx, please review the following functions to see if you can learn what you need there:

* [pygfx.materials.material_from_trimesh](https://github.com/pygfx/pygfx/blob/f163bc82a2d2b244d5aac4500143b7a0f76b34d3/pygfx/materials/_compat.py#L43)
* [pygfx.geometries.geometry_from_trimesh](https://github.com/pygfx/pygfx/blob/f163bc82a2d2b244d5aac4500143b7a0f76b34d3/pygfx/geometries/_compat.py#L6)
* [pygfx.load_meshes](https://github.com/pygfx/pygfx/blob/f163bc82a2d2b244d5aac4500143b7a0f76b34d3/pygfx/utils/load.py#L35)

There is indeed a conversion step required in UV coordinates which you can see in `geometry_from_trimesh` as referenced above.

**Reply by schlegelp** [reply]
Thanks, that's perfect!

**Reply by schlegelp** [reply]
I had to modify the built-in functions slightly to make them work with some of the trimesh objects/files I played with. 

It's all really simple stuff. For example, `CesiumMilkTruck.glb` uses a texture encoded through a palette (`mode='P'`) and unless you explicitly convert to RGB you currently get a greyscale texture in `pygfx`. 

Would you be interested in a PR with these minor fixes?

**Reply by Korijn** [reply]
Yes, absolutely! We aren't heavy users of trimesh ourselves, so this would be super helpful.

**Reply by schlegelp** [reply]
See #707
